src/main/java/ch/njol/skript/sections/EffSecFireworkLaunch.java

```diff
@@ -0,0 +1,166 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+
+package ch.njol.skript.sections;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.config.SectionNode;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.lang.EffectSection;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.Trigger;
+import ch.njol.skript.lang.TriggerItem;
+import ch.njol.skript.registrations.EventValues;
+import ch.njol.skript.util.Getter;
+import ch.njol.skript.variables.Variables;
+import ch.njol.util.Kleenean;
+import org.bukkit.FireworkEffect;
+import org.bukkit.Location;
+import org.bukkit.entity.Firework;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.bukkit.inventory.meta.FireworkMeta;
+import org.bukkit.util.Consumer;
+import org.eclipse.jdt.annotation.Nullable;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+@Name("Launch Firework")
+@Description({"Launch firework effects at the given location(s).",
+	"This can be used as an effect and as a section.",
+	"If it is used as a section, the section is run before the entity is added to the world.",
+	"You can modify the firework in this section, using for example 'event-entity' or 'event-firework'.",
+	"Do note that other event values, such as 'player', won't work in this section."})
+@Examples({"launch ball large coloured red, purple and white fading to light green and black at player's location with duration 1",
+	"",
+	"#Firework launch section example",
+	"on damage:",
+	"\tif damage cause is entity explosion:",
+	"\t\tif metadata value \"cancelDamage\" of event-projectile is true:",
+	"\t\t\tcancel event",
+	"",
+	"command /firework:",
+	"\ttrigger:",
+	"\t\tlaunch a firework with effects ball large coloured red at player:",
+	"\t\t\tset metadata value \"cancelDamage\" of event-firework to true"})
+@Since("2.4, INSERT VERSION (with section)")
+public class EffSecFireworkLaunch extends EffectSection {
+
+	public static class FireworkLaunchEvent extends Event {
+		private final Firework firework;
+
+		public FireworkLaunchEvent(Firework firework) {
+			this.firework = firework;
+		}
+
+		public Firework getFirework() {
+			return firework;
+		}
+
+		@Override
+		public @NotNull
+		HandlerList getHandlers() {
+			throw new IllegalStateException();
+		}
+	}
+
+	static {
+		Skript.registerSection(EffSecFireworkLaunch.class, "(launch|deploy) [[a] firework [with effect[s]]] %fireworkeffects% at %locations% [([with] (duration|power)|timed) %number%]");
```

Comment:
The `%number%` is not null in this case. It should be `%-number%` because when the user doesn't specify a power number we want to handle it later in the `get` method.